src/main/java/org/broadinstitute/hellbender/engine/MultiFeatureWalker.java

```diff
@@ -0,0 +1,246 @@
+package org.broadinstitute.hellbender.engine;
+
+import htsjdk.samtools.SAMSequenceDictionary;
+import htsjdk.samtools.SAMSequenceRecord;
+import htsjdk.tribble.Feature;
+import htsjdk.variant.vcf.VCFHeader;
+import org.broadinstitute.hellbender.cmdline.StandardArgumentDefinitions;
+import org.broadinstitute.hellbender.exceptions.UserException;
+import org.broadinstitute.hellbender.utils.IntervalUtils;
+import org.broadinstitute.hellbender.utils.SimpleInterval;
+import org.broadinstitute.hellbender.tools.sv.SVFeaturesHeader;
+
+import java.util.*;
+
+/**
+ * A MultiFeatureWalker is a tool that presents one {@link Feature} at a time in sorted order from
+ * multiple sources of Features.  The input files for each feature must be sorted by locus.
+ *
+ * To use this walker you need only implement the abstract apply method in a class that declares
+ * a collection of FeatureInputs as an argument.
+ */
+public abstract class MultiFeatureWalker<F extends Feature> extends WalkerBase {
+
+    SAMSequenceDictionary dictionary;
+    final Set<String> samples = new TreeSet<>();
+
+    @Override
+    public boolean requiresFeatures(){
+        return true;
+    }
+
+    @Override
+    public String getProgressMeterRecordLabel() { return "features"; }
+
+    @Override
+    void initializeFeatures() {
+        features = new FeatureManager(this, FeatureDataSource.DEFAULT_QUERY_LOOKAHEAD_BASES,
+                            cloudPrefetchBuffer, cloudIndexPrefetchBuffer, getGenomicsDBOptions());
+    }
+
+    /**
+     * Operations performed just prior to the start of traversal.
+     */
+    @Override
+    public void onTraversalStart() {
+        setDictionaryAndSamples();
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * Implementation of Feature-based traversal.
+     *
+     * NOTE: You should only override {@link #traverse()} if you are writing a new walker base class
+     * in the engine package that extends this class. It is not meant to be overridden by tools
+     * outside the engine package.
+     */
+    @Override
+    public void traverse() {
+        final MergingIterator<F> iterator =
+                new MergingIterator<>(dictionary, features, userIntervals);
+        while ( iterator.hasNext() ) {
+            final PQEntry<F> entry = iterator.next();
+            final F feature = entry.getFeature();
+            apply(feature, entry.getHeader());
+            progressMeter.update(feature);
+        }
+    }
+
+    /**
+     * Process an individual feature.
+     * In general, subclasses should simply stream their output from apply(), and maintain as little
+     * internal state as possible.
+     *
+     * @param feature Current Feature being processed.
+     * @param header Header object for the source from which the feature was drawn (may be null)
+     */
+    public abstract void apply( final F feature, final Object header );
```

Comment:
@tedsharpe GATK has always been structured around universal arguments like `-R` and `-I` that all tools accept, and that (when provided) automatically populate contextual objects passed in to the tools. The main justification for this design is to not require any extra work for the tool authors when overlapping data from other sources is required, and to have the most common kinds of inputs "wired up" in advance and available for the tool to use as needed. Tools can call the `hasReference()`, etc., methods from `GATKTool`, or the `hasBackingDataSource()` methods on the context objects themselves if they need to query whether a particular kind of contextual data is available.

Reply:
Yeah.  I get that.  IMHO, it's not really extra work to get the objects you need when you actually need them, and that's better than preparing dummy objects that you might not even want or use that get passed to you every time.  I know that's how it's always worked, but it seems silly to me.  I revised the code to do things the standard way.

Reply:
It's a little more discoverable / self-documenting this way -- a reminder that the engine can provide you with all these side inputs if you need them. Plus there is no overhead to having these contextual objects around: they are all implemented in terms of lazy queries that don't happen until you access them.

Reply:
Nope.  Comes from the Ministry of Silly Walkers.  That's my story, and I'm stickin' to it.

Reply:
Hah, well, it may not convince you, but here's one last defense: the context objects are all initialized with the interval of the primary record -- since they are "tied" to the primary record in that sense, it makes sense that they should be passed in during the traversal along with the primary record.